src/Microsoft.Win32.Registry/src/Microsoft/Win32/RegistryKey.Windows.cs

```diff
@@ -393,36 +394,47 @@ private int InternalSubKeyCountCore()
             return subkeys;
         }
 
-        private unsafe string[] InternalGetSubKeyNamesCore(int subkeys)
+        private unsafe string[] InternalGetSubKeyNamesCore()
         {
-            string[] names = new string[subkeys];
-            char[] name = new char[MaxKeyLength + 1];
+            List<string> names = new List<string>();
+            char[] name = ArrayPool<char>.Shared.Rent(MaxKeyLength + 1);
 
-            int namelen;
-
-            fixed (char* namePtr = &name[0])
+            try
             {
-                for (int i = 0; i < subkeys; i++)
+                int index = 0;
+                int result;
+                int nameLength = name.Length;
+
+                while ((result = Interop.Advapi32.RegEnumKeyEx(
+                    _hkey,
+                    index,
+                    name,
+                    ref nameLength,
+                    null,
+                    null,
+                    null,
+                    null)) != Interop.Errors.ERROR_NO_MORE_ITEMS)
                 {
-                    namelen = name.Length; // Don't remove this. The API's doesn't work if this is not properly initialized.
-                    int ret = Interop.Advapi32.RegEnumKeyEx(_hkey,
-                        i,
-                        namePtr,
-                        ref namelen,
-                        null,
-                        null,
-                        null,
-                        null);
-                    if (ret != 0)
+                    switch (result)
                     {
-                        Win32Error(ret, null);
+                        case Interop.Errors.ERROR_SUCCESS:
+                            names.Add(new string(name, 0, nameLength));
+                            nameLength = name.Length;
+                            index++;
+                            break;
+                        default:
+                            // Throw the error
+                            Win32Error(result, null);
```

Comment:
In this case I don't see that it would help enough to warrant adding the assert as the fail path is a "normal" codepath that will give the needed follow-up info. We got an error that we didn't expect and we threw- if we were counting on a specific state and falling over incidentally (null ref, etc.) I'd absolutely want the assert.